### ISSUE TITLE
Try running contrast-agnostic model on the EPI data

### ISSUE THREAD

**Issue by jcohenadad** [sct-pipeline/fmri-segmentation#23]
If results look OK, I suggest re-training the contrast-agnostic model so it can also work for EPI data.

**Comment by rohanbanerjee** [comment]
I have run the contrast-agnostic model on the EPI data and visually the results look okay. Use the following command for further reference:


```python
python /home/GRAMES.POLYMTL.CA/robana/duke/temp/rohan/fmri_sc_seg/monai/run_inference_single_image.py --path-img {image_path} --chkp-path /home/GRAMES.POLYMTL.CA/robana/duke/temp/muena/contrast-agnostic/final_monai_model/nnunet_nf=32_DS=1_opt=adam_lr=0.001_AdapW_CCrop_bs=2_64x192x320_20230918-2253 --path-out /home/GRAMES.POLYMTL.CA/robana/duke/temp/rohan/fmri_sc_seg/results/monai_results --device cpu"
```

**Comment by rohanbanerjee** [comment]
I am attaching the qc reports of both ground truths and predictions of the contrast-agnostic model on the test set of the EPI data.

[gt_qc.zip](https://github.com/sct-pipeline/fmri-segmentation/files/12910162/gt_qc.zip)
[monai_preds_qc.zip](https://github.com/sct-pipeline/fmri-segmentation/files/12910163/monai_preds_qc.zip)

**Comment by jcohenadad** [comment]
@rohanbanerjee: 
- please merge QCs into the same QC-- please talk with @Nilser3 he knows exactly how to do it--
- can you please report your conclusions? it is important to also bring your insights/interpretation every time you present results

**Comment by rohanbanerjee** [comment]
[fmri_sc-seg_qc.zip](https://github.com/sct-pipeline/fmri-segmentation/files/13031529/fmri_sc-seg_qc.zip)

### Test performance (Dice score and Hausdorff Distance):
| Model | Dice Score | Hausdorff Distance |
|--------|--------|--------|
| Contrast-agnostic model | 0.84 ± 0.055 | 3.21 ± 1.67 |
| nnUNet | **92.5 ± 0.03** | **1.73 ± 0.84** |
### Interpreting the QC
Each of the test subjects contains 3 different masks. If the value of the `-s` flag contains:
1. `SUBJECT_NAME.nii.gz` --> means ground truth (drawn by the site)
2. `SUBJECT_NAME_pred.nii.gz` --> means predicted mask by the nnUNet model specifically trained for EPI data
3. `SUBJECT_NAME_bin.nii.gz` --> means predicted by the contrast-agnostic model 

### Conclusions regarding the QC (common observations):
All the conclusions do take into account the fact that this dataset is completely out-of-distribution (hasn't seen this type of data during the training) for the `contrast-agnostic` model and the `nnUNet` was trained only on the EPI data.
1. In cases where the spinal cord is clearly visible (i.e. the contrast in between the spinal cord and the CSF is good), the `contrast-agnostic` model performs better or equal to the `nnUNet` model.

2. In cases where it has a not so clear contrast, the `nnUNet` model performs better than the `contrast-agnostic` model i.e. the `contrast-agnostic` model is prone to over/under-segmentation. For example, `sub-nwM10`:

<img width="118" alt="Screenshot 2023-10-18 at 3 00 14 PM" src="https://github.com/sct-pipeline/fmri-segmentation/assets/25586344/212c603c-56d0-4ae4-a6a5-7e8efe5abdd8">

<img width="118" alt="Screenshot 2023-10-18 at 3 01 26 PM" src="https://github.com/sct-pipeline/fmri-segmentation/assets/25586344/346c5be4-0344-4509-9d13-efcf43817c2f">

<img width="116" alt="Screenshot 2023-10-18 at 3 03 19 PM" src="https://github.com/sct-pipeline/fmri-segmentation/assets/25586344/770c46b2-e1dd-40a9-9378-b1689a929070">

<img width="119" alt="Screenshot 2023-10-18 at 3 33 11 PM" src="https://github.com/sct-pipeline/fmri-segmentation/assets/25586344/c7a79596-139e-48e1-82ef-2ed8b6a0e844">

(order: Image --> ground truth --> `nnUNet` --> `contrast-agnostic`)

Next step would be to fine-tune the `contrast-agnostic` model and compare with baselines.

**Comment by rohanbanerjee** [comment]
Relevant issue: https://github.com/sct-pipeline/contrast-agnostic-softseg-spinalcord/issues/83

**Comment by rohanbanerjee** [comment]
Updating the comment with what has been tried until now. 

## Fine-tuning:
### Objective
The main objective of the fine-tuning was to use the contrast-agnostic model pre-trained weights to transfer knowledge to a model which is trained to segment spinal cord on EPI data. This is how the newly trained model would be agnostic to EPI data. 

It is expected that the fine-tuned model would have a good performance since it would have a lot of "spinal cord" context through its weights and biases.

Path to checkpoint used: `duke/temp/muena/contrast-agnostic/final_monai_model/nnunet_nf=32_DS=1_opt=adam_lr=0.001_AdapW_CCrop_bs=2_64x192x320_20230918-2253`

This PR https://github.com/sct-pipeline/contrast-agnostic-softseg-spinalcord/pull/85 adds the functionality to initialize a model with pre-trained weights. 

### Results and Observations
1. **Test Dice Score:** 0.47 ± 0.24; **Test Hausdorff Distance:** 9.69 ± 7.97
2. Unstable validation soft dice curve:
<img width="500" alt="Screenshot 2023-11-08 at 12 53 21 PM" src="https://github.com/sct-pipeline/contrast-agnostic-softseg-spinalcord/assets/25586344/0136e751-bfee-4868-a039-620522064ec5">

After investigation, the reason of this poor result and unstable training was the crop size. The crop size used for the training was `64x192x320`. This is also what the `contrast-agnostic` model was trained on. BUT, the image size of the EPI data is smaller than the crop size of the `contrast-agnostic` model. The median crop size of the EPI data (as per nnUNet plans) should be `32x48x128`. If this crop size is used then the model throws out an `error` since there is a `size mismatch` with the pre-trained model. This is the end feeds noise to the model which ends up in the unstable training.

### Conclusion:
1. Seems like the `contrast-agnostic` model cannot be trained out-of-the-box because of the `size mismatch` error. 
2. The workaround can be resize the image by padding.

**Comment by jcohenadad** [comment]
If you set `CCrop_bs=2_64x192x320`, then doesn't MONAI do some padding on the input image?

**Comment by rohanbanerjee** [comment]
We have decided to continue with nnUNet for now
